<p align="right"><a href="./C1_W3_Lab05_Cost_Function_Soln.ipynb">English</a> | <strong>中文</strong></p>

# Optional Lab: 逻辑回归的代价函数（Cost Function）

## 目标
在本实验中，你将：
- 考察并使用逻辑回归的代价函数。

In [1]:
import numpy as np
%matplotlib widget
import matplotlib.pyplot as plt
from lab_utils_common import  plot_data, sigmoid, dlc
plt.style.use('./deeplearning.mplstyle')

## 数据集
我们从决策边界实验用过的同一个数据集开始。

In [2]:
X_train = np.array([[0.5, 1.5], [1,1], [1.5, 0.5], [3, 0.5], [2, 2], [1, 2.5]])  #(m,n)
y_train = np.array([0, 0, 0, 1, 1, 1])                                           #(m,)

我们用一个辅助函数把数据画出来。标签 $y=1$ 的数据点显示为红色叉号，标签 $y=0$ 的数据点显示为蓝色圆圈。

In [3]:
fig,ax = plt.subplots(1,1,figsize=(4,4))
plot_data(X_train, y_train, ax)

# Set both axes to be from 0-4
ax.axis([0, 4, 0, 3.5])
ax.set_ylabel('$x_1$', fontsize=12)
ax.set_xlabel('$x_0$', fontsize=12)
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

## 代价函数

在前一个实验里，你推导了 *逻辑损失（logistic loss）* 函数。回忆一下，loss 针对单个样本定义。这里你把这些 loss 合起来构成 **cost（代价）**，它包含所有样本。

回忆一下，逻辑回归的代价函数形如

$$ J(\mathbf{w},b) = \frac{1}{m} \sum_{i=0}^{m-1} \left[ loss(f_{\mathbf{w},b}(\mathbf{x}^{(i)}), y^{(i)}) \right] \tag{1}$$

其中
* $loss(f_{\mathbf{w},b}(\mathbf{x}^{(i)}), y^{(i)})$ 是单个数据点的代价，即：

    $$loss(f_{\mathbf{w},b}(\mathbf{x}^{(i)}), y^{(i)}) = -y^{(i)} \log\left(f_{\mathbf{w},b}\left( \mathbf{x}^{(i)} \right) \right) - \left( 1 - y^{(i)}\right) \log \left( 1 - f_{\mathbf{w},b}\left( \mathbf{x}^{(i)} \right) \right) \tag{2}$$

*  其中 m 是数据集中的训练样本数量，且：
$$
\begin{align}
  f_{\mathbf{w},b}(\mathbf{x^{(i)}}) &= g(z^{(i)})\tag{3} \\
  z^{(i)} &= \mathbf{w} \cdot \mathbf{x}^{(i)}+ b\tag{4} \\
  g(z^{(i)}) &= \frac{1}{1+e^{-z^{(i)}}}\tag{5}
\end{align}
$$

<a name='ex-02'></a>
#### 代码说明

`compute_cost_logistic` 的算法遍历所有样本，计算每个样本的 loss 并累加成总和。

注意，变量 X 和 y 不是标量，而是形状分别为 ($m, n$) 和 ($𝑚$,) 的矩阵，其中 $𝑛$ 是特征数量，$𝑚$ 是训练样本数量。

In [4]:
def compute_cost_logistic(X, y, w, b):
    """
    Computes cost

    Args:
      X (ndarray (m,n)): Data, m examples with n features
      y (ndarray (m,)) : target values
      w (ndarray (n,)) : model parameters  
      b (scalar)       : model parameter
      
    Returns:
      cost (scalar): cost
    """

    m = X.shape[0]
    cost = 0.0
    for i in range(m):
        z_i = np.dot(X[i],w) + b
        f_wb_i = sigmoid(z_i)
        cost +=  -y[i]*np.log(f_wb_i) - (1-y[i])*np.log(1-f_wb_i)
             
    cost = cost / m
    return cost


用下面的 cell 检查代价函数的实现。

In [6]:
w_tmp = np.array([1,1])
b_tmp = -3
print(compute_cost_logistic(X_train, y_train, w_tmp, b_tmp))

0.36686678640551745


**期望输出**：0.3668667864055175

## 示例
现在，我们看看代价函数对不同的 $w$ 值输出什么。

* 在前一个实验里，你画出了 $b = -3, w_0 = 1, w_1 = 1$ 的决策边界，也就是 `b = -3, w = np.array([1,1])`。

* 假设你想看看 $b = -4, w_0 = 1, w_1 = 1$（即 `b = -4, w = np.array([1,1])`）是否是更好的模型。

我们先把这两个不同 $b$ 值的决策边界画出来，看哪个更好地拟合数据。

* 对 $b = -3, w_0 = 1, w_1 = 1$，画 $-3 + x_0+x_1 = 0$（蓝色）
* 对 $b = -4, w_0 = 1, w_1 = 1$，画 $-4 + x_0+x_1 = 0$（品红色）

In [7]:
import matplotlib.pyplot as plt

# Choose values between 0 and 6
x0 = np.arange(0,6)

# Plot the two decision boundaries
x1 = 3 - x0
x1_other = 4 - x0

fig,ax = plt.subplots(1, 1, figsize=(4,4))
# Plot the decision boundary
ax.plot(x0,x1, c=dlc["dlblue"], label="$b$=-3")
ax.plot(x0,x1_other, c=dlc["dlmagenta"], label="$b$=-4")
ax.axis([0, 4, 0, 4])

# Plot the original data
plot_data(X_train,y_train,ax)
ax.axis([0, 4, 0, 4])
ax.set_ylabel('$x_1$', fontsize=12)
ax.set_xlabel('$x_0$', fontsize=12)
plt.legend(loc="upper right")
plt.title("Decision Boundary")
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

从这张图可以看出 `b = -4, w = np.array([1,1])` 对训练数据是更差的模型。我们看看代价函数的实现是否反映了这一点。

In [ ]:
w_array1 = np.array([1,1])
b_1 = -3
w_array2 = np.array([1,1])
b_2 = -4

print("Cost for b = -3 : ", compute_cost_logistic(X_train, y_train, w_array1, b_1))
print("Cost for b = -4 : ", compute_cost_logistic(X_train, y_train, w_array2, b_2))

**期望输出**

b = -3 的代价 Cost：0.3668667864055175

b = -4 的代价 Cost：0.5036808636748461

可以看到代价函数表现符合预期，`b = -4, w = np.array([1,1])` 的代价确实高于 `b = -3, w = np.array([1,1])` 的代价。

## 恭喜！
在本实验中，你考察并使用了逻辑回归的代价函数。